pkg/controllers/provisioningv2/rke2/machineprovision/controller.go

```diff
@@ -623,29 +646,33 @@ func constructFilesSecret(driver string, config map[string]interface{}) (*corev1
 	return nil, nil
 }
 
-func shouldForceRemoval(job *batchv1.Job, d data.Object) bool {
-	var createJobCondition summary.Condition
-	for _, c := range summary.GetUnstructuredConditions(d) {
-		if c.Type() == createJobConditionType {
-			createJobCondition = c
-			break
-		}
+func shouldCleanupObjects(job *batchv1.Job, d data.Object) bool {
+	if !job.Status.CompletionTime.IsZero() {
+		return true
+	}
+
+	createJobCondition := getCondition(d, createJobConditionType)
+	if createJobCondition == nil {
+		return false
 	}
 
 	forceRemoveAnnValue, _ := d.Map("metadata", "annotations")[forceRemoveMachineAnn].(string)
 
-	if createJobCondition.Reason() != string(errors.CreateMachineError) &&
-		strings.ToLower(forceRemoveAnnValue) != "true" {
-		return false
+	if createJobCondition.Reason() == string(errors.CreateMachineError) || strings.ToLower(forceRemoveAnnValue) == "true" {
+		return strings.ToLower(job.Spec.Template.Labels[InfraJobRemove]) == "true" && condition.Cond("Failed").IsTrue(job)
 	}
 
-	return condition.Cond("Failed").IsTrue(job)
+	return false
 }
 
 func FailedMachineDeleteMessage(kind, failureReason, failureMessage string) string {
 	return fmt.Sprintf("failed deleting server (%s) in infrastructure provider: %s: %s", kind, failureReason, failureMessage)
 }
 
+func FailedMachineCreateMessage(kind, failureReason, failureMessage string) string {
```

Comment:
Though why we don't make this `func` unexported, but I see the others like this are used in `rancher/src/github.com/rancher/rancher/pkg/controllers/provisioningv2/rke2/machinestatus/machinestatus.go` - do we need to add handling of this one there too?

Reply:
It's exported because the other, related, functions are also exported. It doesn't need to be used in the spot you mentioned because the status will already be put on the machine status. If this changes, then we will have to manually do it in the same way you mentioned.

Reply:
If you feel strongly that it should be unexported, I can do that. Unless you object, I will keep it as is.

Reply:
"As is" is fine, I just wanted to make sure we didn't need to use it in that spot where the other 3 are used. Resolving.